File: updater.py
#!/usr/bin/env python3

import json
import os
import re
import signal
import subprocess
import sys
import threading
from dataclasses import dataclass
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

# Constants
GITHUB_API_URL = "https://api.github.com/repos/Voxelum/x-minecraft-launcher/releases/latest"
PKGBUILD_FILE = "PKGBUILD"
SRCINFO_FILE = ".SRCINFO"
HTTP_TIMEOUT = 15
BUILD_TIMEOUT = 600  # 10 minutes
USER_AGENT = "xmcl-package-updater/2.0"

# Colors
COLOR_RESET = "\033[0m"
COLOR_RED = "\033[91m"
COLOR_GREEN = "\033[92m"
COLOR_YELLOW = "\033[93m"
COLOR_BLUE = "\033[94m"


class Logger:
    """Provides colored console output"""

    @staticmethod
    def success(msg, *args):
        print(f"{COLOR_GREEN}✓ {msg % args}{COLOR_RESET}")

    @staticmethod
    def error(msg, *args):
        print(f"{COLOR_RED}✗ {msg % args}{COLOR_RESET}", file=sys.stderr)

    @staticmethod
    def info(msg, *args):
        print(f"{COLOR_YELLOW}ℹ {msg % args}{COLOR_RESET}")

    @staticmethod
    def action(msg, *args):
        print(f"{COLOR_BLUE}→ {msg % args}{COLOR_RESET}")


log = Logger()


@dataclass
class Config:
    """Application configuration"""

    http_timeout: int = HTTP_TIMEOUT
    build_timeout: int = BUILD_TIMEOUT


class HTTPClient:
    """Simple HTTP client using urllib"""

    def __init__(self, timeout):
        self.timeout = timeout

    def get_json(self, url):
        req = Request(url, headers={"User-Agent": USER_AGENT})
        try:
            with urlopen(req, timeout=self.timeout) as response:
                if response.status != 200:
                    raise Exception(f"HTTP {response.status}")
                data = response.read().decode("utf-8")
                return json.loads(data)
        except HTTPError as e:
            raise Exception(f"HTTP {e.code}: {e.reason}")
        except URLError as e:
            raise Exception(f"URL error: {e.reason}")


class FileManager:
    """Handles file operations"""

    @staticmethod
    def exists(path):
        return Path(path).exists()

    @staticmethod
    def read_lines(path):
        try:
            with open(path, "r") as f:
                return f.read().split("\n")
        except Exception as e:
            raise Exception(f"reading file {path}: {e}")

    @staticmethod
    def write_lines(path, lines):
        try:
            with open(path, "w") as f:
                f.write("\n".join(lines))
        except Exception as e:
            raise Exception(f"writing file {path}: {e}")


class CommandRunner:
    """Executes system commands"""

    @staticmethod
    def run(name, *args):
        proc = subprocess.Popen([name] + list(args), stdout=sys.stdout, stderr=sys.stderr)
        returncode = proc.wait()
        if returncode != 0:
            raise Exception(f"running {name}: exit code {returncode}")

    @staticmethod
    def run_with_output(name, *args):
        proc = subprocess.Popen([name] + list(args), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        stdout, stderr = proc.communicate()
        if proc.returncode != 0:
            raise Exception(f"running {name}: {stderr.decode()}")
        return stdout


class PKGBUILDUpdater:
    """Handles PKGBUILD modifications"""

    def __init__(self, fm):
        self.fm = fm

    def update_version(self, version):
        lines = self.fm.read_lines(PKGBUILD_FILE)

        version_num = version.lstrip("v")
        pkgver_regex = re.compile(r"^pkgver=.*$")
        pkgrel_regex = re.compile(r"^pkgrel=.*$")

        for i, line in enumerate(lines):
            trimmed = line.strip()
            if pkgver_regex.match(trimmed):
                lines[i] = f"pkgver={version_num}"
            elif pkgrel_regex.match(trimmed):
                lines[i] = "pkgrel=1"

        self.fm.write_lines(PKGBUILD_FILE, lines)


class ReleaseChecker:
    """Fetches latest version from GitHub"""

    def __init__(self, http_client):
        self.http_client = http_client

    def get_latest_version(self):
        try:
            release = self.http_client.get_json(GITHUB_API_URL)
            tag_name = release.get("tag_name")
            if not tag_name:
                raise Exception("no tag_name in API response")
            return tag_name
        except Exception as e:
            raise Exception(f"fetching latest release: {e}")


class UserInteraction:
    """Handles user prompts"""

    def __init__(self):
        self.cancelled = False

    def confirm_build(self):
        print("Build package now? [y/N]: ", end="", flush=True)

        if self.cancelled:
            return False

        try:
            response = input().strip().lower()
            return response in ("y", "yes")
        except (EOFError, KeyboardInterrupt):
            print()
            return False


class PackageBuilder:
    """Handles package building operations"""

    def __init__(self, cmd_runner, fm):
        self.cmd_runner = cmd_runner
        self.fm = fm
        self.build_process = None
        self.build_lock = threading.Lock()

    def update_checksums(self):
        self.cmd_runner.run("updpkgsums")

    def update_srcinfo(self):
        output = self.cmd_runner.run_with_output("makepkg", "--printsrcinfo")

        try:
            with open(SRCINFO_FILE, "wb") as f:
                f.write(output)
        except Exception as e:
            raise Exception(f"writing {SRCINFO_FILE}: {e}")

    def build_package(self, timeout):
        with self.build_lock:
            self.build_process = subprocess.Popen(
                ["makepkg", "-s"],
                stdout=sys.stdout,
                stderr=sys.stderr,
                preexec_fn=os.setpgrp,  # Create new process group
            )

        try:
            returncode = self.build_process.wait(timeout=timeout)
            if returncode != 0:
                raise Exception(f"makepkg failed with exit code {returncode}")
        except subprocess.TimeoutExpired:
            self.kill_build()
            raise Exception("build timeout")

    def kill_build(self):
        with self.build_lock:
            if self.build_process and self.build_process.poll() is None:
                try:
                    # Kill the process group
                    os.killpg(os.getpgid(self.build_process.pid), signal.SIGTERM)
                except (ProcessLookupError, PermissionError):
                    pass


class Application:
    """Main application"""

    def __init__(self, config):
        self.config = config
        http_client = HTTPClient(config.http_timeout)
        self.fm = FileManager()
        cmd_runner = CommandRunner()

        self.release_checker = ReleaseChecker(http_client)
        self.pkgbuild_updater = PKGBUILDUpdater(self.fm)
        self.package_builder = PackageBuilder(cmd_runner, self.fm)
        self.user_interaction = UserInteraction()
        self.cancelled = False

    def validate_environment(self):
        if not self.fm.exists(PKGBUILD_FILE):
            raise Exception(f"{PKGBUILD_FILE} not found in current directory")

        required_commands = ["updpkgsums", "makepkg"]
        for cmd in required_commands:
            if not self._command_exists(cmd):
                raise Exception(f"required command '{cmd}' not found in PATH")

    @staticmethod
    def _command_exists(cmd):
        from shutil import which

        return which(cmd) is not None

    def update_package_files(self, version):
        log.action("Updating PKGBUILD...")
        self.pkgbuild_updater.update_version(version)

        log.action("Updating checksums...")
        self.package_builder.update_checksums()

        log.action("Updating .SRCINFO...")
        self.package_builder.update_srcinfo()

    def run(self):
        # Validate environment
        self.validate_environment()

        # Get latest version
        log.action("Fetching latest version from GitHub...")
        try:
            version = self.release_checker.get_latest_version()
        except Exception as e:
            if self.cancelled:
                raise Exception("operation cancelled")
            raise Exception(f"getting latest version: {e}")

        if self.cancelled:
            raise Exception("operation cancelled")

        log.info("Latest version: %s", version)

        # Update package files
        try:
            self.update_package_files(version)
        except Exception as e:
            if self.cancelled:
                raise Exception("operation cancelled")
            raise e

        if self.cancelled:
            raise Exception("operation cancelled")

        log.success("Package files updated to version %s", version)

        # Ask user about building
        try:
            should_build = self.user_interaction.confirm_build()
        except Exception as e:
            raise Exception(f"getting user input: {e}")

        if self.cancelled:
            raise Exception("operation cancelled")

        if not should_build:
            log.info("Skipping package build")
            return

        # Build package
        log.action("Building package...")
        try:
            self.package_builder.build_package(self.config.build_timeout)
        except Exception as e:
            if self.cancelled:
                raise Exception("build cancelled")
            raise e

        if self.cancelled:
            raise Exception("build cancelled")

        log.success("Package built successfully!")

    def handle_interrupt(self, signum, frame):
        log.info("Interrupt received, cancelling...")
        self.cancelled = True
        self.user_interaction.cancelled = True
        self.package_builder.kill_build()
        sys.exit(1)


def main():
    config = Config()
    app = Application(config)

    # Handle interrupts
    signal.signal(signal.SIGINT, app.handle_interrupt)
    signal.signal(signal.SIGTERM, app.handle_interrupt)

    try:
        app.run()
    except Exception as e:
        log.error("%s", str(e))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: updater.mjs
#!/usr/bin/env bun

const GITHUB_API_URL = "https://api.github.com/repos/Voxelum/x-minecraft-launcher/releases/latest";
const PKGBUILD_FILE = "PKGBUILD";
const SRCINFO_FILE = ".SRCINFO";
const HTTP_TIMEOUT = 15;
const BUILD_TIMEOUT = 600; // 10 minutes
const USER_AGENT = "xmcl-package-updater/2.0";

// Colors
const COLOR_RESET = "\x1b[0m";
const COLOR_RED = "\x1b[91m";
const COLOR_GREEN = "\x1b[92m";
const COLOR_YELLOW = "\x1b[93m";
const COLOR_BLUE = "\x1b[94m";

class Logger {
  static #format(msg, ...args) {
    let i = 0;
    return msg.replace(/%s/g, () => args[i++] ?? "");
  }

  static success(msg, ...args) {
    console.log(`${COLOR_GREEN}✓ ${Logger.#format(msg, ...args)}${COLOR_RESET}`);
  }

  static error(msg, ...args) {
    console.error(`${COLOR_RED}✗ ${Logger.#format(msg, ...args)}${COLOR_RESET}`);
  }

  static info(msg, ...args) {
    console.log(`${COLOR_YELLOW}ℹ ${Logger.#format(msg, ...args)}${COLOR_RESET}`);
  }

  static action(msg, ...args) {
    console.log(`${COLOR_BLUE}→ ${Logger.#format(msg, ...args)}${COLOR_RESET}`);
  }
}

class Config {
  constructor(httpTimeout = HTTP_TIMEOUT, buildTimeout = BUILD_TIMEOUT) {
    this.httpTimeout = httpTimeout;
    this.buildTimeout = buildTimeout;
  }
}

class HTTPClient {
  constructor(timeout) {
    this.timeout = timeout;
  }

  async getJson(url) {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout * 1000);

    try {
      const res = await fetch(url, {
        headers: { "User-Agent": USER_AGENT },
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      return await res.json();
    } catch (err) {
      clearTimeout(timeoutId);
      if (err.name === "AbortError") throw new Error("HTTP timeout");
      throw new Error(err.message);
    }
  }
}

class FileManager {
  exists(path) {
    return Bun.file(path).exists();
  }

  async readLines(path) {
    try {
      const text = await Bun.file(path).text();
      return text.split("\n");
    } catch (err) {
      throw new Error(`reading file ${path}: ${err.message}`);
    }
  }

  async writeLines(path, lines) {
    try {
      await Bun.write(path, lines.join("\n"));
    } catch (err) {
      throw new Error(`writing file ${path}: ${err.message}`);
    }
  }
}

class CommandRunner {
  async run(name, ...args) {
    const proc = Bun.spawn([name, ...args], {
      stdout: "inherit",
      stderr: "inherit",
    });
    const exitCode = await proc.exited;
    if (exitCode !== 0) {
      throw new Error(`running ${name}: exit code ${exitCode}`);
    }
  }

  async runWithOutput(name, ...args) {
    const proc = Bun.spawn([name, ...args], {
      stdout: "pipe",
      stderr: "pipe",
    });

    const [stdout, stderr] = await Promise.all([
      new Response(proc.stdout).text(),
      new Response(proc.stderr).text(),
    ]);

    const exitCode = await proc.exited;
    if (exitCode !== 0) {
      throw new Error(`running ${name}: ${stderr.trim() || "unknown error"}`);
    }
    return stdout;
  }
}

class PKGBUILDUpdater {
  constructor(fm) {
    this.fm = fm;
  }

  async updateVersion(version) {
    let lines = await this.fm.readLines(PKGBUILD_FILE);

    const versionNum = version.replace(/^v/, "");
    const pkgverRegex = /^pkgver=.*$/;
    const pkgrelRegex = /^pkgrel=.*$/;

    for (let i = 0; i < lines.length; i++) {
      const trimmed = lines[i].trim();
      if (pkgverRegex.test(trimmed)) {
        lines[i] = `pkgver=${versionNum}`;
      } else if (pkgrelRegex.test(trimmed)) {
        lines[i] = "pkgrel=1";
      }
    }

    await this.fm.writeLines(PKGBUILD_FILE, lines);
  }
}

class ReleaseChecker {
  constructor(http) {
    this.http = http;
  }

  async getLatestVersion() {
    try {
      const release = await this.http.getJson(GITHUB_API_URL);
      const tagName = release?.tag_name;
      if (!tagName) throw new Error("no tag_name in API response");
      return tagName;
    } catch (err) {
      throw new Error(`fetching latest release: ${err.message}`);
    }
  }
}

class UserInteraction {
  cancelled = false;

  async confirmBuild() {
    if (this.cancelled) return false;

    process.stdout.write("Build package now? [y/N]: ");

    const input = await new Promise((resolve) => {
      const listener = (data) => {
        process.stdin.removeListener("data", listener);
        process.stdin.pause();
        resolve(data.toString().trim().toLowerCase());
      };
      process.stdin.resume();
      process.stdin.once("data", listener);
    });

    return input === "y" || input === "yes";
  }
}

class PackageBuilder {
  buildProcess = null;

  constructor(cmd, fm) {
    this.cmd = cmd;
    this.fm = fm;
  }

  async updateChecksums() {
    await this.cmd.run("updpkgsums");
  }

  async updateSrcinfo() {
    const output = await this.cmd.runWithOutput("makepkg", "--printsrcinfo");
    await Bun.write(SRCINFO_FILE, output);
  }

  async buildPackage(timeout) {
    this.buildProcess = Bun.spawn(["makepkg", "-s"], {
      stdout: "inherit",
      stderr: "inherit",
    });

    const proc = this.buildProcess;

    try {
      await Promise.race([
        proc.exited.then((code) => {
          if (code !== 0) throw new Error(`makepkg failed with exit code ${code}`);
        }),
        new Promise((_, reject) =>
          setTimeout(() => reject(new Error("build timeout")), timeout * 1000),
        ),
      ]);
    } catch (err) {
      if (err.message === "build timeout") this.killBuild();
      throw err;
    }
  }

  killBuild() {
    if (this.buildProcess?.pid) {
      try {
        process.kill(-this.buildProcess.pid, "SIGTERM");
      } catch {}
    }
  }
}

class Application {
  cancelled = false;

  constructor(config) {
    this.config = config;
    this.fm = new FileManager();
    this.cmd = new CommandRunner();

    this.releaseChecker = new ReleaseChecker(new HTTPClient(config.httpTimeout));
    this.pkgbuildUpdater = new PKGBUILDUpdater(this.fm);
    this.packageBuilder = new PackageBuilder(this.cmd, this.fm);
    this.userInteraction = new UserInteraction();
  }

  validateEnvironment() {
    if (!this.fm.exists(PKGBUILD_FILE)) {
      throw new Error(`${PKGBUILD_FILE} not found in current directory`);
    }

    const required = ["updpkgsums", "makepkg"];
    for (const cmd of required) {
      if (!this.#commandExists(cmd)) {
        throw new Error(`required command '${cmd}' not found in PATH`);
      }
    }
  }

  #commandExists(cmd) {
    try {
      const proc = Bun.spawnSync(["sh", "-c", `command -v "${cmd}"`]);
      return proc.exitCode === 0;
    } catch {
      return false;
    }
  }

  async updatePackageFiles(version) {
    Logger.action("Updating PKGBUILD...");
    await this.pkgbuildUpdater.updateVersion(version);

    Logger.action("Updating checksums...");
    await this.packageBuilder.updateChecksums();

    Logger.action("Updating .SRCINFO...");
    await this.packageBuilder.updateSrcinfo();
  }

  async run() {
    this.validateEnvironment();

    Logger.action("Fetching latest version from GitHub...");
    const version = await this.releaseChecker.getLatestVersion();

    Logger.info("Latest version: %s", version);

    await this.updatePackageFiles(version);
    Logger.success("Package files updated to version %s", version);

    const shouldBuild = await this.userInteraction.confirmBuild();
    if (!shouldBuild) {
      Logger.info("Skipping package build");
      return;
    }

    Logger.action("Building package...");
    await this.packageBuilder.buildPackage(this.config.buildTimeout);
    Logger.success("Package built successfully!");
  }

  handleInterrupt() {
    Logger.info("Interrupt received, cancelling...");
    this.cancelled = true;
    this.userInteraction.cancelled = true;
    this.packageBuilder.killBuild();
    process.exit(1);
  }
}

async function main() {
  const config = new Config();
  const app = new Application(config);

  process.on("SIGINT", () => app.handleInterrupt());
  process.on("SIGTERM", () => app.handleInterrupt());

  try {
    await app.run();
  } catch (err) {
    Logger.error("%s", err.message);
    process.exit(1);
  }
}

main();
